Use new openWindow functions in ui, small fixes

cullenn opened the pull request:
- Updated old calls directly to state.openWindow to new window state functions.
- Made friends list icon toggle ui instead of only creating it
- Added saveState() to new window state functions.
- Updated openWindow functions to alert console if new openWindow keys were being added instead of just updating existing keys since there's no other way to be alerted by errors like setWindowClosed('friendsListzzzzz');
- Updated state to compare against a stateTemplate and clean openWindow keys that aren't explicitly defined by us already
- Updated openWindows key names since we're not expecting to need unique names for proxies
- Added merchant to openWindows keys in state

timebomb0 commented on src/mods/friendsList/index.js:
```diff
@@ -16,10 +13,10 @@ function customFriendsList() {
 
 	// Create the friends list UI
-	document.querySelector('.js-friends-list-icon').addEventListener('click', createFriendsList);
+	document.querySelector('.js-friends-list-icon').addEventListener('click', toggleFriendsList);
```
Nice change! Having all the logic in the `ui` and a `windowState` makes this so much easier than it originally would've been. Cool to see.

timebomb0 commented on src/utils/ui.js:
```diff
@@ -239,13 +243,33 @@ function createXpMeter() {
+	}
+
 	state.openWindows[windowName] = true;
+	saveState();
```
Totally forgot to save the state here, thanks for adding this.

cullenn commented:
Oh hell yeah, that's a way better approach to it, thanks! I kept them camelcase and starting with lowercase but let me know if you'd rather that following your example for constants like these. I've committed the updates and listed them below:

- Added in windowNames constant and updated references to prevent window key typos.
- Removed extra logic previously added to state.js for state.openWindow key cleanup.
- Removed alerts for new keys being added in ui.js window state functions.

cullenn commented:
Also was curious, any preference with how many functions are being imported to a file before switching to import *? With friendsList/index.js it's importing 4 functions form ui.js, but with player.js it's using 6 functions from ui.js and just running import *

timebomb0 commented:
I'm pretty loose on `import *` vs `import {...}` thus far. Generally speaking, I care more about how the consuming code looks than the import statements.

My current philosophy is: If the method sounds like it is unrelated to the mod's code, then using `import * as myUtil` can help the code be arguably slightly more readable, e.g. `chat.addChatMessage` in the `itemStatsCopy` mod.

That being said, it's a pretty loose, biased philosophy, and I haven't been completely consistent with it in the code base.

Feel free to do whatever you think is best with that in mind. If I feel strongly enough, I'll mention my opinion for specific instances :)

timebomb0 commented on src/utils/ui.js:
```diff
@@ -2,6 +2,13 @@ import { getState, saveState } from './state';
 import { makeElement } from './misc';
 import * as player from './player';
 
+let windowNames = {
```
1) Can we rename this to `WindowNames`? I'm a fan of differentiating constant variables, i.e. literal strings/ints/etc set in the code and never changed (unlike a lot of our instances of `const varName = ...`), naming from other variables. Some codebases also use the `WINDOW_NAMES` naming standard, but that's slightly more of a hassle to type, so I'm fine with `WindowNames`.

2) This can be `const` :)

cullenn replied:
Sounds good, done!

timebomb0 commented:
Everything else LGTM - will test this tomorrow sometime and merge it in once the final piece of feedback s addressed. Cheers and happy New Years!

cullenn commented:
Feedback addressed, all yours!

timebomb0 commented:
Looks good to me, just merged it in, thanks for the PR!
